pkg/app/piped/imagewatcher/watcher.go

```diff
@@ -13,68 +13,187 @@
 // limitations under the License.
 
 // Package imagewatcher provides a piped component
-// that periodically checks the image registry and updates
+// that periodically checks the container registry and updates
 // the image if there are differences with Git.
 package imagewatcher
 
 import (
 	"context"
+	"sync"
 	"time"
 
 	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/app/piped/imageprovider"
+	"github.com/pipe-cd/pipe/pkg/config"
+	"github.com/pipe-cd/pipe/pkg/git"
 )
 
 type Watcher interface {
 	Run(context.Context) error
 }
 
-type watcher struct {
-	timer  *time.Timer
-	logger *zap.Logger
+type gitClient interface {
+	Clone(ctx context.Context, repoID, remote, branch, destination string) (git.Repo, error)
 }
 
-type imageRepos map[string]imageRepo
-type imageRepo struct {
+type watcher struct {
+	config    *config.PipedSpec
+	gitClient gitClient
+	logger    *zap.Logger
+	wg        sync.WaitGroup
+	mu        sync.Mutex
+
+	// Indexed by repo id.
+	gitRepos sync.Map
```

Comment:
Not sure why do we need to use a sync map instead of a normal map?

Reply:
Oops, you're right. A map can be read when it's being read. Gonna fix them.